java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java

```diff
@@ -0,0 +1,196 @@
+  JavaType applySubstitution(JavaType type, JavaType site) {
+    if (isParametrizedType(site)) {
+      return applySubstitution(type, ((JavaType.ParametrizedTypeJavaType) site).typeSubstitution);
+    }
+    return type;
+  }
+
+  List<JavaType> applySubstitution(List<JavaType> types, TypeSubstitution substitution) {
+    if (substitution.size() == 0) {
+      return types;
+    }
+    List<JavaType> results = new ArrayList<>(types.size());
+    for (JavaType type : types) {
+      results.add(applySubstitution(type, substitution));
+    }
+    return results;
+  }
+
+  private JavaType applySubstitution(JavaType type, TypeSubstitution substitution) {
+    JavaType substitutedType = substitution.substitutedType(type);
+    if (substitutedType != null) {
+      return substitutedType;
+    }
+    if (isParametrizedType(type)) {
+      return applySubstitution((JavaType.ParametrizedTypeJavaType) type, substitution);
+    }
+    if (isWildcardType(type)) {
+      return applySubstitution((JavaType.WildCardType) type, substitution);
+    }
+    if (isArrayType(type)) {
+      return applySubstitution((JavaType.ArrayJavaType) type, substitution);
+    }
+    return type;
+  }
+
+  private static boolean isParametrizedType(JavaType type) {
+    return type instanceof JavaType.ParametrizedTypeJavaType;
+  }
+
+  private static boolean isWildcardType(JavaType type) {
+    return type instanceof JavaType.WildCardType;
+  }
+
+  private static boolean isArrayType(JavaType type) {
+    return type instanceof JavaType.ArrayJavaType;
+  }
+
+  private JavaType applySubstitution(JavaType.ParametrizedTypeJavaType type, TypeSubstitution substitution) {
+    TypeSubstitution newSubstitution = new TypeSubstitution();
+    for (Map.Entry<JavaType.TypeVariableJavaType, JavaType> entry : type.typeSubstitution.substitutionEntries()) {
+      newSubstitution.add(entry.getKey(), applySubstitution(entry.getValue(), substitution));
+    }
+    return parametrizedTypeCache.getParametrizedTypeType(type.rawType.getSymbol(), newSubstitution);
+  }
+
+  private JavaType applySubstitution(JavaType.WildCardType wildcard, TypeSubstitution substitution) {
+    JavaType substitutedType = substitution.substitutedType(wildcard.bound);
+    if (substitutedType != null) {
+      return parametrizedTypeCache.getWildcardType(substitutedType, wildcard.boundType);
+    }
+    return wildcard;
+  }
+
+  private JavaType applySubstitution(JavaType.ArrayJavaType arrayType, TypeSubstitution substitution) {
+    JavaType rootElementType = arrayType.elementType;
+    int nbDimensions = 1;
+    while (rootElementType.isTagged(JavaType.ARRAY)) {
+      rootElementType = ((JavaType.ArrayJavaType) rootElementType).elementType;
+      nbDimensions++;
+    }
+    JavaType substitutedType = substitution.substitutedType(rootElementType);
+    if (substitutedType != null) {
+      // FIXME SONARJAVA-1574 a new array type should not be created but reused if already existing for the current element type
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)